File: backend/app/routers/animals.py
"""
Animals router: CRUD operations for animal records (farm-scoped).
"""

from fastapi import APIRouter, Depends, HTTPException, Query, Request, status
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.db_models import Animal, User, utcnow
from app.models.schemas import AnimalCreate, AnimalResponse, AnimalUpdate, Page
from app.services.audit_service import record_audit
from app.services.auth_service import get_current_user, require_manager
from app.utils import integrity_guard

_TAG_TAKEN = "Tag number is already assigned to another animal"

router = APIRouter(prefix="/api/animals", tags=["animals"])


def _get_owned_animal(db: Session, animal_id: int, user: User) -> Animal:
    """Fetch an animal within the caller's farm or raise 404."""
    animal = (
        db.query(Animal)
        .filter(Animal.id == animal_id, Animal.farm_id == user.farm_id)
        .first()
    )
    if not animal:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Animal not found"
        )
    return animal


@router.get("", response_model=Page[AnimalResponse])
def list_animals(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
    page: int = Query(1, ge=1),
    limit: int = Query(50, ge=1, le=200),
):
    """Return this farm's animal records, most recent first (paginated)."""
    base = db.query(Animal).filter(Animal.farm_id == current_user.farm_id)
    total = base.count()
    items = (
        base.order_by(Animal.created_at.desc())
        .offset((page - 1) * limit)
        .limit(limit)
        .all()
    )
    return Page(items=items, total=total, page=page, limit=limit)


@router.post("", response_model=AnimalResponse, status_code=status.HTTP_201_CREATED)
def create_animal(
    request: Request,
    body: AnimalCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Create a new animal record in the caller's farm."""
    if body.tag_number:
        existing = (
            db.query(Animal)
            .filter(
                Animal.farm_id == current_user.farm_id,
                Animal.tag_number == body.tag_number,
            )
            .first()
        )
        if existing:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=_TAG_TAKEN)
    animal = Animal(
        farm_id=current_user.farm_id,
        name=body.name,
        animal_type=body.animal_type,
        tag_number=body.tag_number,
        breed=body.breed,
        date_of_birth=body.date_of_birth,
        notes=body.notes,
        added_by_user_id=current_user.id,
        status="alive",
    )
    db.add(animal)
    with integrity_guard(db, _TAG_TAKEN):
        db.commit()
    db.refresh(animal)
    record_audit(
        db,
        request,
        current_user,
        action="create",
        entity_type="animal",
        entity_id=animal.id,
        details={"name": animal.name, "tag_number": animal.tag_number},
    )
    return animal


@router.get("/{animal_id}", response_model=AnimalResponse)
def get_animal(
    animal_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Return a single animal record from the caller's farm."""
    return _get_owned_animal(db, animal_id, current_user)


@router.put("/{animal_id}", response_model=AnimalResponse)
def update_animal(
    request: Request,
    animal_id: int,
    body: AnimalUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Update an animal record.

    Both roles can update; only managers can force-set status.
    """
    animal = _get_owned_animal(db, animal_id, current_user)

    if body.status == "dead" and current_user.role != "manager":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail=(
                "Only managers can directly set status"
                " to 'dead'. Use the death report endpoint."
            ),
        )

    update_data = body.model_dump(exclude_unset=True)

    new_tag = update_data.get("tag_number")
    if new_tag:
        clash = (
            db.query(Animal)
            .filter(
                Animal.farm_id == current_user.farm_id,
                Animal.tag_number == new_tag,
                Animal.id != animal_id,
            )
            .first()
        )
        if clash:
            raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=_TAG_TAKEN)

    for field, value in update_data.items():
        setattr(animal, field, value)
    animal.updated_at = utcnow()

    with integrity_guard(db, _TAG_TAKEN):
        db.commit()
    db.refresh(animal)
    record_audit(
        db,
        request,
        current_user,
        action="update",
        entity_type="animal",
        entity_id=animal.id,
        details=update_data,
    )
    return animal


@router.delete("/{animal_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_animal(
    request: Request,
    animal_id: int,
    db: Session = Depends(get_db),
    current_manager: User = Depends(require_manager),
):
    """Delete an animal record. Manager access required."""
    animal = _get_owned_animal(db, animal_id, current_manager)
    db.delete(animal)
    db.commit()
    record_audit(
        db,
        request,
        current_manager,
        action="delete",
        entity_type="animal",
        entity_id=animal_id,
    )

File: supabase/functions/api/routers/animals.ts
// Port of backend/app/routers/animals.py.
import { Hono } from "hono";
import { getDb, isUniqueViolation } from "../../_shared/db.ts";
import { parsePagination } from "../../_shared/pagination.ts";
import {
  csvResponse,
  jsonError,
  pageEnvelope,
} from "../../_shared/response.ts";
import { recordAudit } from "../../_shared/auditService.ts";
import { requireAuth, requireManager } from "../../_shared/auth.ts";
import {
  AnimalCreateSchema,
  AnimalUpdateSchema,
  firstZodError,
} from "../../_shared/validation.ts";

const TAG_TAKEN = "Tag number is already assigned to another animal";
const NOT_FOUND = "Animal not found";

export const animalsRouter = new Hono();

// Every route in this router requires an authenticated user; there is no
// public animals endpoint (matches app/routers/animals.py, which depends on
// get_current_user for all routes and require_manager additionally for delete).
animalsRouter.use("*", requireAuth);

interface AnimalRow {
  id: number;
  farm_id: number;
  name: string;
  animal_type: string;
  tag_number: string | null;
  breed: string | null;
  // Cast to text in every query below — postgres.js otherwise returns `date`
  // columns as JS Date objects at UTC midnight, which JSON.stringify renders
  // as a full "...T00:00:00.000Z" timestamp instead of Pydantic's plain
  // "YYYY-MM-DD" string. Casting keeps the wire format identical to the
  // FastAPI original and avoids timezone-shift bugs.
  date_of_birth: string | null;
  status: "alive" | "dead";
  notes: string | null;
  added_by_user_id: number;
  created_at: Date;
  updated_at: Date;
}

const UPDATABLE_FIELDS = [
  "name",
  "animal_type",
  "tag_number",
  "breed",
  "date_of_birth",
  "status",
  "notes",
] as const;

function parseAnimalId(raw: string): number | null {
  if (!/^\d+$/.test(raw)) return null;
  return parseInt(raw, 10);
}

/** Fetch an animal within the caller's farm, or null (mirrors _get_owned_animal). */
async function getOwnedAnimal(
  animalId: number,
  farmId: number,
): Promise<AnimalRow | null> {
  const sql = getDb();
  const [animal] = await sql<AnimalRow[]>`
    select id, farm_id, name, animal_type, tag_number, breed,
           date_of_birth::text as date_of_birth, status, notes,
           added_by_user_id, created_at, updated_at
    from animals
    where id = ${animalId} and farm_id = ${farmId}
  `;
  return animal ?? null;
}

animalsRouter.get("/", async (c) => {
  const user = c.get("user");
  const paged = parsePagination(c);
  if (!paged.ok) return jsonError(c, 422, paged.error);
  const { page, limit, offset } = paged.pagination;

  const sql = getDb();
  const [{ count }] = await sql<{ count: string }[]>`
    select count(*) as count from animals where farm_id = ${user.farm_id}
  `;
  const items = await sql<AnimalRow[]>`
    select id, farm_id, name, animal_type, tag_number, breed,
           date_of_birth::text as date_of_birth, status, notes,
           added_by_user_id, created_at, updated_at
    from animals
    where farm_id = ${user.farm_id}
    order by created_at desc
    offset ${offset} limit ${limit}
  `;
  return c.json(pageEnvelope(items, parseInt(count, 10), page, limit));
});

animalsRouter.post("/", async (c) => {
  const user = c.get("user");
  const body = await c.req.json().catch(() => null);
  const parsed = AnimalCreateSchema.safeParse(body);
  if (!parsed.success) return jsonError(c, 422, firstZodError(parsed));
  const data = parsed.data;

  const sql = getDb();

  // App-level pre-check for a friendlier error in the common (non-racy) case;
  // the per-farm unique constraint (uq_animal_farm_tag) is the source of
  // truth and is still guarded below via isUniqueViolation.
  if (data.tag_number) {
    const [existing] = await sql<{ id: number }[]>`
      select id from animals
      where farm_id = ${user.farm_id} and tag_number = ${data.tag_number}
    `;
    if (existing) return jsonError(c, 409, TAG_TAKEN);
  }

  let animal: AnimalRow;
  try {
    [animal] = await sql<AnimalRow[]>`
      insert into animals (
        farm_id, name, animal_type, tag_number, breed, date_of_birth, notes,
        added_by_user_id, status
      )
      values (
        ${user.farm_id}, ${data.name}, ${data.animal_type},
        ${data.tag_number ?? null}, ${data.breed ?? null},
        ${data.date_of_birth ?? null}, ${data.notes ?? null},
        ${user.id}, 'alive'
      )
      returning id, farm_id, name, animal_type, tag_number, breed,
                date_of_birth::text as date_of_birth, status, notes,
                added_by_user_id, created_at, updated_at
    `;
  } catch (err) {
    if (isUniqueViolation(err)) return jsonError(c, 409, TAG_TAKEN);
    throw err;
  }

  await recordAudit({
    c,
    actor: user,
    action: "create",
    entityType: "animal",
    entityId: animal.id,
    details: { name: animal.name, tag_number: animal.tag_number },
  });

  return c.json(animal, 201);
});

animalsRouter.get("/export.csv", async (c) => {
  const user = c.get("user");
  const sql = getDb();
  const animals = await sql<AnimalRow[]>`
    select id, name, animal_type, tag_number, breed,
           date_of_birth::text as date_of_birth, status, notes, created_at
    from animals
    where farm_id = ${user.farm_id}
    order by created_at desc
  `;
  const header = [
    "id",
    "name",
    "animal_type",
    "tag_number",
    "breed",
    "date_of_birth",
    "status",
    "notes",
    "created_at",
  ];
  const rows = animals.map((a) => [
    a.id,
    a.name,
    a.animal_type,
    a.tag_number ?? "",
    a.breed ?? "",
    a.date_of_birth ?? "",
    a.status,
    a.notes ?? "",
    // .toISOString() rather than Python's bare str(datetime): the raw JS
    // Date's default toString() ("Mon Jan 01 2024 ...") has no Python
    // equivalent worth matching, so we emit a stable, parseable timestamp.
    a.created_at.toISOString(),
  ]);
  return csvResponse(c, "animals.csv", header, rows);
});

animalsRouter.get("/:id", async (c) => {
  const user = c.get("user");
  const id = parseAnimalId(c.req.param("id"));
  if (id === null) {
    return jsonError(c, 422, "animal_id: Input should be a valid integer");
  }
  const animal = await getOwnedAnimal(id, user.farm_id);
  if (!animal) return jsonError(c, 404, NOT_FOUND);
  return c.json(animal);
});

animalsRouter.put("/:id", async (c) => {
  const user = c.get("user");
  const id = parseAnimalId(c.req.param("id"));
  if (id === null) {
    return jsonError(c, 422, "animal_id: Input should be a valid integer");
  }

  const existing = await getOwnedAnimal(id, user.farm_id);
  if (!existing) return jsonError(c, 404, NOT_FOUND);

  const raw = await c.req.json().catch(() => null);
  if (raw === null || typeof raw !== "object") {
    return jsonError(c, 422, "Invalid request body");
  }
  const parsed = AnimalUpdateSchema.safeParse(raw);
  if (!parsed.success) return jsonError(c, 422, firstZodError(parsed));

  if (parsed.data.status === "dead" && user.role !== "manager") {
    return jsonError(
      c,
      403,
      "Only managers can directly set status to 'dead'. Use the death report endpoint.",
    );
  }

  // Mirror Pydantic's model_dump(exclude_unset=True): only fields explicitly
  // present in the request body are applied — fields the client omitted must
  // leave the stored value untouched, even though Zod fills in `optional()`
  // fields as undefined either way.
  const rawObject = raw as Record<string, unknown>;
  const updateData: Record<string, unknown> = {};
  for (const field of UPDATABLE_FIELDS) {
    if (Object.prototype.hasOwnProperty.call(rawObject, field)) {
      updateData[field] = (parsed.data as Record<string, unknown>)[field];
    }
  }

  const sql = getDb();

  if (updateData.tag_number) {
    const [clash] = await sql<{ id: number }[]>`
      select id from animals
      where farm_id = ${user.farm_id}
        and tag_number = ${updateData.tag_number as string}
        and id != ${id}
    `;
    if (clash) return jsonError(c, 409, TAG_TAKEN);
  }

  let updated: AnimalRow;
  try {
    const columns = Object.keys(updateData);
    if (columns.length > 0) {
      [updated] = await sql<AnimalRow[]>`
        update animals set ${sql(updateData, ...columns)}, updated_at = now()
        where id = ${id} and farm_id = ${user.farm_id}
        returning id, farm_id, name, animal_type, tag_number, breed,
                  date_of_birth::text as date_of_birth, status, notes,
                  added_by_user_id, created_at, updated_at
      `;
    } else {
      // No recognized fields were present in the body — still bump
      // updated_at, matching the Python handler which always sets
      // animal.updated_at = utcnow() before committing, even on a no-op body.
      [updated] = await sql<AnimalRow[]>`
        update animals set updated_at = now()
        where id = ${id} and farm_id = ${user.farm_id}
        returning id, farm_id, name, animal_type, tag_number, breed,
                  date_of_birth::text as date_of_birth, status, notes,
                  added_by_user_id, created_at, updated_at
      `;
    }
  } catch (err) {
    if (isUniqueViolation(err)) return jsonError(c, 409, TAG_TAKEN);
    throw err;
  }

  await recordAudit({
    c,
    actor: user,
    action: "update",
    entityType: "animal",
    entityId: updated.id,
    details: updateData,
  });

  return c.json(updated);
});

animalsRouter.delete("/:id", requireManager, async (c) => {
  const user = c.get("user");
  const id = parseAnimalId(c.req.param("id"));
  if (id === null) {
    return jsonError(c, 422, "animal_id: Input should be a valid integer");
  }

  const existing = await getOwnedAnimal(id, user.farm_id);
  if (!existing) return jsonError(c, 404, NOT_FOUND);

  const sql = getDb();
  await sql`delete from animals where id = ${id} and farm_id = ${user.farm_id}`;

  await recordAudit({
    c,
    actor: user,
    action: "delete",
    entityType: "animal",
    entityId: id,
  });

  return c.body(null, 204);
});
